zkevm-circuits/src/state_circuit/binary_number.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{
+    evm_circuit::{
+        table::RwTableTag,
+        util::{and, not},
+    },
+    util::Expr,
+};
+use eth_types::Field;
+use halo2_proofs::{
+    circuit::Region,
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Fixed, VirtualCells},
+    poly::Rotation,
+};
+use std::collections::BTreeSet;
+use std::marker::PhantomData;
+use strum::IntoEnumIterator;
+
+pub trait AsBits<const N: usize> {
+    // Return the bits of self, starting from the most significant.
+    fn as_bits(&self) -> [bool; N];
+}
+
+impl AsBits<4> for RwTableTag {
+    fn as_bits(&self) -> [bool; 4] {
+        let mut bits = [false; 4];
+        let mut x = *self as u8;
+        for i in 0..4 {
+            bits[3 - i] = x % 2 == 1;
+            x /= 2;
+        }
+        bits
+    }
+}
+
+impl<const N: usize> AsBits<N> for usize {
+    fn as_bits(&self) -> [bool; N] {
+        let mut bits = [false; N];
+        let mut x = *self as u64;
+        for i in 0..N {
+            bits[N - 1 - i] = x % 2 == 1;
+            x /= 2;
+        }
+        bits
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Config<T, const N: usize>
+where
+    T: AsBits<N>,
+{
+    pub bits: [Column<Advice>; N],
+    _marker: PhantomData<T>,
+}
+
+impl<T, const N: usize> Config<T, N>
+where
+    T: AsBits<N>,
+{
+    pub fn value<F: Field>(
+        &self,
+        rotation: Rotation,
+    ) -> impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F> + '_ {
+        move |meta: &mut VirtualCells<'_, F>| {
+            let bits = self.bits.map(|bit| meta.query_advice(bit, rotation));
+            bits.iter()
+                .fold(0.expr(), |result, bit| bit.clone() + result * 2.expr())
+        }
+    }
+
+    pub fn value_equals<F: Field, S: AsBits<N>>(
+        &self,
+        value: &S,
+        rotation: Rotation,
+    ) -> impl FnOnce(&mut VirtualCells<'_, F>) -> Expression<F> + '_ {
+        let bits = value.as_bits();
+        move |meta: &mut VirtualCells<'_, F>| {
+            Self::value_equals_expr(
+                &bits,
+                &self.bits.map(|bit| meta.query_advice(bit, rotation)),
+            )
+        }
+    }
+
+    pub fn value_equals_expr<F: Field>(
+        value: &[bool],
+        expressions: &[Expression<F>],
+    ) -> Expression<F> {
+        and::expr(value.iter().zip(expressions).map(|(&bit, expression)| {
+            if bit {
+                expression.clone()
+            } else {
+                not::expr(expression.clone())
+            }
+        }))
+    }
+}
+
+pub struct Chip<F: Field, T, const N: usize>
```

Comment:
Could you add some documentation saying what `T` and `N` are?

Reply:
Suggestion for documentation:
This chip helps working with binary encoding of integers of length `N` bits by:
- Enforcing that the binary representation is in the valid range defined by `T`.
- Creating expressions (via the `Config`) that evaluate to `1` when the bits match a specific value.

Reply:
thanks. I've added these.